src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs

```diff
@@ -1649,8 +1649,22 @@ public sealed class Options : BinaryOptionsBase
             /// <value>
             /// If unspecified, <see cref="LogLoss"/> will be used.
             /// </value>
-            [Argument(ArgumentType.Multiple, HelpText = "Loss Function", ShortName = "loss", SortOrder = 50)]
-            public ISupportSdcaClassificationLossFactory LossFunction = new LogLossFactory();
+            [Argument(ArgumentType.Multiple, Name = "LossFunction", HelpText = "Loss Function", ShortName = "loss", SortOrder = 50)]
+            internal ISupportSdcaClassificationLossFactory LossFunctionFactory = new LogLossFactory();
+
+            /// <summary>
+            /// The custom <a href="tmpurl_loss">loss</a>.
+            /// </summary>
+            /// <value>
+            /// If unspecified, <see cref="LogLoss"/> will be used.
+            /// </value>
+            public ISupportSdcaClassificationLoss LossFunction;
```

Comment:
>LossFunction [](start = 50, length = 12)

It doesn't look like we're using the same pattern Tom mentioned this morning. Please take a look at the following example for hiding `EarlyStoppingRuleFactory` via adding a new `EarlyStoppingRule`.
```csharp
        /// <summary>
        /// Early stopping rule. (Validation set (/valid) is required).
        /// </summary>
        [BestFriend]
        [Argument(ArgumentType.Multiple, HelpText = "Early stopping rule. (Validation set (/valid) is required.)", ShortName = "esr", NullName = "<Disable>")]
        [TGUI(Label = "Early Stopping Rule", Description = "Early stopping rule. (Validation set (/valid) is required.)")]
        internal IEarlyStoppingCriterionFactory EarlyStoppingRuleFactory;

        private EarlyStoppingRuleBase _earlyStoppingRuleBase;

        public EarlyStoppingRuleBase EarlyStoppingRule
        {
            get { return _earlyStoppingRuleBase;  }
            set
            {
                _earlyStoppingRuleBase = value;
                EarlyStoppingRuleFactory = _earlyStoppingRuleBase.BuildFactory();
            }
        }
```

Reply:
I dont see benefits of the changes you are proposing. What does _earlyStoppingRuleBase.BuildFactory(); do? How does it know what factory to build out if I set non default EarlyStoppingRule ?

---
In reply to: [262721736](https://github.com/dotnet/machinelearning/pull/2856#discussion_r262721736) [](ancestors = 262721736)

Reply:
One benefit is that the way Wei-Sheng proposes means that the object does not really enter any sort of inconsistent state, even for our internal code, thereby making the object easier to maintain. We could of course say, since the only thing touching this object at that level is this code, that we can avoid any flaws, but -- let's be honest with ourselves, even internal inconsistencies often show up as bugs externally. :) If you don't want to do it right now @ganik that is fine with me, but I would prefer in that case if we could make it a property so that we could adjust it later, if we believe it is necessary, in subsequent versions, without it being a breaking change. (So we'd change, here, `public ISupportSdcaClassificationLoss LossFunction { get; set; }` so that it becomes an auto-implemented property.

---
In reply to: [262729182](https://github.com/dotnet/machinelearning/pull/2856#discussion_r262729182) [](ancestors = 262729182,262721736)

Reply:
Just going to re-activate for tracking. Like I say, you don't have to do it right now, but I'd rather it be a property in this case.

---
In reply to: [263112293](https://github.com/dotnet/machinelearning/pull/2856#discussion_r263112293) [](ancestors = 263112293,262729182,262721736)

Reply:
Sure I am fine with properties. For me the importance was that as third party dev I can supply my own LossFunction definition and will not be restricted to few pre-defined loss functions: LogLoss, HingeLoss etc. With Wei-Sheng pattern this becomes prohibitive.

---
In reply to: [263112433](https://github.com/dotnet/machinelearning/pull/2856#discussion_r263112433) [](ancestors = 263112433,263112293,262729182,262721736)

Reply:
Modified to properties

---
In reply to: [263631467](https://github.com/dotnet/machinelearning/pull/2856#discussion_r263631467) [](ancestors = 263631467,263112433,263112293,262729182,262721736)